### PR TITLE
[RFR] Test Embedded Ansible integration for Nuage

### PR DESCRIPTION
**Pull request by aljazkosir**
Test that Nuage events trigger Embedded Ansible automation and that playbook has access to authentication attributes and event data.

Specifically, we copy AE Instance 'ManageIQ/System/Event/EmsEvent/Nuage/nuage_enterprise_create' from default domain into our own domain and customize it's 'meth5' attribute to invoke Embedded Ansible playbook which prints authentication attributes and event data into evm.log. This test then triggers a 'nuage_enterprise_create' event and waits for appropriate line to appear in evm.log.

Prerequisites:
Following content needs to be present in cfme_data.yaml in order to fetch correct Ansible repository:

```yaml
# cfme_data.yaml
ansible_links:
  playbook_repositories:
    embedded_ansible: https://github.com/xlab-si/integration-tests-nuage-automation.git
```

Video explaining what the test is doing: 
https://drive.google.com/file/d/1weHHLjQ_0t1OrXECaDDxoGjzjYwwJ18R/view

\cc @miha-plesko

### CODE REVIEW

File: cfme/fixtures/embedded_ansible.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    destination_domain = appliance.collections.domains.instantiate(destination_domain_name)
+    destination_namespace = create_namespaces(destination_domain, source)
+    return destination_namespace.classes.instantiate(source_class_name)\
+        .instances.instantiate(source_instance_name)
+
+
+@pytest.fixture
+def domain(request, appliance):
+    """
+    Created new domain with random name, if we want we can also provide a custom name with
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: Be careful, first line in docstring is special so you shouldn't break it:

```python

"""
Create a new AE Domain.

By default, a random name is generated for the Domain, but you can specify it explicitly with @pytest.mark.parametrize
decorator on your test function.

Example:
@pytest.mark.parametrize('domain', ['TestDomain'], indirect=True)
def test_my_automation(domain):
  ...
"""
```

File: cfme/fixtures/embedded_ansible.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    @pytest.mark.parametrize test decorator.
+
+    Example: @pytest.mark.parametrize('domain', ['TestDomain'], indirect=True)
+    """
+    dc = appliance.collections.domains
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: What does `dc` stand for? I think we should just do

```python
domain = appliance.collections.domains.create(...)
```

When possible, I'd suggest we use full names instead short-hand names (`domain` vs `d`)

File: cfme/fixtures/embedded_ansible.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    d = dc.create(
+        name=getattr(request, 'param', None) or random_name(),
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]:
```python
name=getattr(request, 'param', random_name()),
```

File: cfme/fixtures/embedded_ansible.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        enabled=True)
+    yield d
+    d.delete()
+
+
+@pytest.fixture
+def copy_ae_instance_to_domain_name(request, appliance):
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: Since we don't make use of this fixture ATM I'd just remove it. It's simple enough to implement it later if needed.

File: cfme/fixtures/embedded_ansible.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    [('ManageIQ/System/Event/EmsEvent/Nuage/nuage_enterprise_create', 'TestDomain')], indirect=True)
+
+    :return: copied AE instance with provided existing domain
+    """
+    src, dest = request.param
+    return copy_ae_instance(appliance, src, dest)
+
+
+@pytest.fixture
+def copy_ae_instance_to_domain(request, appliance, domain):
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: Please rename to `copy_ae_instance_to_new_domain`

File: cfme/fixtures/embedded_ansible.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        .instances.instantiate(source_instance_name).copy_to(destination_domain_name)
+
+    destination_domain = appliance.collections.domains.instantiate(destination_domain_name)
+    destination_namespace = create_namespaces(destination_domain, source)
+    return destination_namespace.classes.instantiate(source_class_name)\
+        .instances.instantiate(source_instance_name)
+
+
+@pytest.fixture
+def domain(request, appliance):
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: Please rename to `custom_ae_domain`

File: cfme/fixtures/embedded_ansible.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    This is going to create new domain and copy AE instance to it. We can also choose name of the
+    new domain by using one more decorator, see domain fixture docstring
+
+    Example:
+
+    @pytest.mark.parametrize('copy_ae_instance_to_domain',
+    ['ManageIQ/System/Event/EmsEvent/Nuage/nuage_enterprise_create'], indirect=True)
+
+    :return: copied AE instance with new domain
+    """
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: """
    Copy AE Instance to a newly created AE Domain.

    This fixture requires at least Instance fqdn parameter, while newly created domain name
    can also be customized:

    @pytest.mark.parametrize(
      'copy_ae_instance_to_new_domain',
      ['ManageIQ/System/Event/EmsEvent/Nuage/nuage_enterprise_create'],
      indirect=True
    )
    @pytest.mark.parametrize(
      'custom_ae_domain',
      ['CustomDomainName'],
      indirect=True
    )
    def test_my_automation(copy_ae_instance_to_domain):
      ...

    :return: newly created AE Instance
    """

File: cfme/fixtures/embedded_ansible.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    src = request.param
+    return copy_ae_instance(appliance, src, domain.name)
```
Comment (miha-plesko) [inline cfme/fixtures/embedded_ansible.py]: Nah, just make it one-liner:

```python
return copy_ae_instance(appliance, request.param, custom_ae_domain.name)
```

File: cfme/tests/networks/nuage/test_automation.py [hunk 9]
```diff
@@ -12,3 +22,52 @@ def test_auth_attributes_of_nuage_ae_class(appliance):
...
+    nuage_ae_class = copy_ae_instance_to_domain.klass
+    ae_method = nuage_ae_class.methods.create(name="printout", location='playbook',
+                                              repository=ansible_repository.name,
+                                              playbook='printout.yaml',
+                                              machine_credential="CFME Default Credential",
+                                              logging_output='Always')
+    username = random_name()
+    with update(copy_ae_instance_to_domain):
+        copy_ae_instance_to_domain.fields = {
+            'nuage_username': {'value': username},
```
[the remark this answers is not in the document]
Comment (aljazkosir) [reply cfme/tests/networks/nuage/test_automation.py]: There is actually a problem with password field, because password field has a different `id` value (`id='inst_password_value'`) inside HTML tag then the rest of the fields (`id='inst_value')`. File `cfme/automate/explorer/instance.py` line 117 and 124.

\cc @miha-plesko

**Comment by aljazkosir** [comment]
I included most of the changes, issue with password mentioned above still exists and parametrizing cfme_data with custom url.

\cc @miha-plesko

**Comment by dajoRH** [comment]
Lint report for commit 895c323fcd161959a127c83075d9ee9129611590:

`cfme/fixtures/embedded_ansible.py`:
- :bangbang: Line 42:101: **E501** *line too long (122 > 100 characters)*

`cfme/tests/networks/nuage/test_automation.py`:
- :heavy_exclamation_mark: Line 91:73: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by aljazkosir** [comment]
I rebased and added parametrization to ansible playbook url.

\cc @miha-plesko

File: cfme/tests/networks/nuage/test_automation.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (miha-plesko) [reply cfme/tests/networks/nuage/test_automation.py]: Yaaay. Okay, I see you managed to resolve this by simply not including the password, which works with me 👍

**Comment by aljazkosir** [comment]
@mshriver As requested I edited the pull request with the video explaining what the test is doing.

**Comment by miha-plesko** [comment]
Thanks @aljazkosir I think this PR is now ready to go

File: cfme/automate/explorer/method.py [hunk 10]
```diff
@@ -433,7 +434,7 @@ def tree_path(self):
     def create(
             self, name=None, display_name=None, location='inline', script=None, data=None,
             cancel=False, validate=True, repository=None, playbook=None, machine_credential=None,
-            hosts=None, max_ttl=None, escalate_privilege=None, verbosity=None,
+            hosts=None, max_ttl=None, logging_output=None, escalate_privilege=None, verbosity=None,
```
Comment (mshriver) [inline cfme/automate/explorer/method.py]: This attribute should be on the `Method` entity, and passed to the instantiate call as well.

File: cfme/fixtures/ansible_fixtures.py [hunk 11]
```diff
@@ -18,12 +18,25 @@ def wait_for_ansible(appliance):
 
 
 @pytest.fixture(scope="module")
-def ansible_repository(appliance, wait_for_ansible):
+def ansible_repository(request, appliance, wait_for_ansible):
+    """
+    By default cfme_data.ansible_links.playbook_repositories.embedded_ansible is set for the url,
```
Comment (mshriver) [inline cfme/fixtures/ansible_fixtures.py]: :+1: Great enhancement

File: cfme/fixtures/embedded_ansible.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def create_namespaces(d, namespace_names):
+    namespace = None
+    for namespace_name in namespace_names:
+        if namespace:
+            namespace = namespace.namespaces.instantiate(namespace_name)
+        else:
+            namespace = d.namespaces.instantiate(namespace_name)
+    return namespace
```
[the remark this answers is not in the document]
Comment (mshriver) [reply cfme/fixtures/embedded_ansible.py]: Yes, I think we need some docblock context here in what the function is actually doing here.

Its odd to me, at a read, why only the last namespace instantiated is returned, when multiple are being instantiated.

File: cfme/fixtures/embedded_ansible.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    :param domain_name: name of exisitng AE Domain where we want to copy to e.g. 'TestDomain'
+    :return: the new AE Instance
+    """
+    source = instance_fqdn.split('/')
+    source_domain_name = source.pop(0)
+    source_instance_name = source.pop(-1)
+    source_class_name = source.pop(-1)
+    source_domain = appliance.collections.domains.instantiate(source_domain_name)
+    namespace = open_namespace(source_domain, source)
+    namespace.classes.instantiate(source_class_name)\
```
Comment (mshriver) [inline cfme/fixtures/embedded_ansible.py]: Please do not use `\` line continuation, wrap with parenthesis or change syntax of calls.

Also this call chain is not clear - please include a comment in-line in plain english of what resource you're trying to use and copy

File: cfme/fixtures/embedded_ansible.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+def open_namespace(domain_name, namespace_levels):
+    namespace = domain_name
+    for namespace_name in namespace_levels:
+        namespace = namespace.namespaces.instantiate(namespace_name)
+    return namespace
+
+
+def copy_ae_instance(appliance, instance_fqdn, domain_name):
+    """
+    Copy AE Instance into a domain.
```
Comment (mshriver) [inline cfme/fixtures/embedded_ansible.py]: This seems like a complex method to have within the fixtures module, instead of as a method on Instance. 

I see that the `copy_to` method that you're using on the AE Instance, defined here, has a `TODO` for 'Namespace!', which is actually not clear to me what needs to be done there.

https://github.com/ManageIQ/integration_tests/blob/c94723bf2bb7aebd66a1e5e8130fb76f2ff935c2/cfme/automate/explorer/common.py#L30

Can this `Copiable.copy_to` method be expanded to cover your use case here?

Otherwise, can you provide more description of exactly what is going on in this method - there are calls to pop components off of the instance fqdn that aren't clear, and then chained calls to instantiate entities within collections that are not commented and not clear.

File: cfme/automate/explorer/method.py (hunk 10 above)
Comment (mshriver) [reply cfme/automate/explorer/method.py]: @aljazkosir This is required - the Entity for this collection doesn't have this attribute. It should be stored with the entity object, otherwise its invisible to the framework outside of this create method, where its passed to the form fill.

File: cfme/fixtures/embedded_ansible.py (hunk 14 above)
Comment (aljazkosir) [reply cfme/fixtures/embedded_ansible.py]: `Copiable.copy_to` requires instance of an object, but the way that test decoration works, we can't provide AE instance to a fixture. We instead provide a full string path to AE instance, proper classes are then generated inside this function.

I split the call chains into multiple variables so the call is more clear and also provided some explanation and an example.

File: cfme/fixtures/embedded_ansible.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    'ManageIQ' as source_domain
+    'nuage_enterprise_create' as source AE instance
+    'Nuage' as source AE class
+    the remaining strings left after 'pop' calls are the AE namespaces which are instantiated
+    through 'instantiate_namespace' function
+    AE instance is then copied to a new domain. We also instantiate all the destination classes so
+    we get proper parent-child structure.
+
+
+    :param appliance: appliance
```
Comment (mshriver) [inline cfme/fixtures/embedded_ansible.py]: Thanks so much for adding this docblock content, this is great and lays out your intent very well.

Please use sphynx-napoleon style for these docblocks.
https://cfme-tests.readthedocs.io/en/master/guides/documenting.html?highlight=napoleon

This section should look more like:
```
Args:
    appliance: Appliance object
    instance_fqdn (str): full path <yada>
```

Comment (aljazkosir) [reply cfme/fixtures/embedded_ansible.py]: Updated :)

File: cfme/tests/networks/nuage/test_automation.py [hunk 16]
```diff
@@ -12,3 +22,74 @@ def test_auth_attributes_of_nuage_ae_class(appliance):
...
     assert all(expected in schema_fields for expected in expected_fields)
+
+
+@pytest.mark.parametrize(
+    'copy_ae_instance_to_new_domain',
+    ['ManageIQ/System/Event/EmsEvent/Nuage/nuage_enterprise_create'],
+    indirect=True)
+@pytest.mark.parametrize(
+    'ansible_repository',
+    [cfme_data.ansible_links.playbook_repositories.nuage],
```
Comment (mshriver) [inline cfme/tests/networks/nuage/test_automation.py]: This breaks collection of tests when this entry isn't present in the cfme_data.yaml that is available.  Don't want to have a direct cfme_data reference here where it can cause an exception.

Consider parametrizing the playbook name (nuage), and have the fixture wrap the lookup in try/except to skip the test when that playbook repo name isn't available.

```
================ ERRORS ================
________ ERROR at setup of cfme/tests/networks/nuage/test_automation.py ___________
cfme/tests/networks/nuage/test_automation.py:33: in <module>
    [cfme_data.ansible_links.playbook_repositories.nuage],
.py2env/lib/python2.7/site-packages/lya/lya.py:68: in __getattr__
    if not (k.startswith('__') or k.startswith('_OrderedDict__')): return self[k]
E   KeyError: 'nuage'
```

Comment (aljazkosir) [reply cfme/tests/networks/nuage/test_automation.py]: I changed how we pass the playbook into the fixture.

**Comment by aljazkosir** [comment]
@mshriver as requested I?m attaching cfme log for this test:
[cfme.log](https://github.com/ManageIQ/integration_tests/files/3028430/cfme.log)

**Comment by dajoRH** [comment]
Lint report for commit 7e0bab75fca9325e19ad928fda826b17cc8f3952:

`cfme/tests/networks/nuage/test_automation.py`:
- :red_circle: Line 5:1: **F401** *'cfme.utils.conf.cfme_data' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/fixtures/ansible_fixtures.py [hunk 17]
```diff
@@ -18,12 +18,23 @@ def wait_for_ansible(appliance):
...
+    but you can specify it explicitly with @pytest.mark.parametrize decorator on your test function.
+
+    Example:
+    @pytest.mark.parametrize('ansible_repository', ['nuage'], indirect=True)
+    def test_function(ansible_repository):
+        ...
+    """
     repositories = appliance.collections.ansible_repositories
+    playbooks = cfme_data.ansible_links.playbook_repositories
```
Comment (mshriver) [inline cfme/fixtures/ansible_fixtures.py]: I'm going to push a commit tweaking how this is referenced.

Its important that this attribute lookup chain is within the try block below, because a cfme_data file without `ansible_links` or `playbook_repositories` maps will cause an exception during the fixture execution.
